# Customer Transaction Dataset Data Cleaning
I created an ecommerce customer transaction dataset using multiple datasets on an eCommerce platform. This company wants to use the new dataset to analyze products and customers to support their marketing campaigns. After creating the dataset, I am comfronted with cleaning the data to  ake it more useful for the maketing teams ad campaign.

As a demonstration, I use the [Brazilian E-Commerce Public Dataset by Olist on Kaggle](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce). These dataset were selected because like the Olist is an eCommerce platform that connects small business with customers from all over Brazil hence has similar characteristis as the platform I carried out the task for. 

Baring in mind that I have already merged all the datasets on the platform into one dataset (customer transaction dataset), I proceed to clean tha data to make sure it is more useful for the marketing teams ad campaigns. So I considered the following:
- working with a large datasets usually create problems including PC memory issues. Hence, these memory utilization issues must be avoided.
- some of the data in the the dataset (e.g. dates) could be 
    
    (a) in the wrong format 
    
    (b) duplicates, 
    
    (c) empty/missing values present
    
- use Pandas `map()` function to map the values in a column of the dataframe to fill out missing values in another column. 

In this simple tutorial, we will look at how to use the map() function to map values in a series to another set of values, both using a custom function and using a mapping from a Python dictionary.
#### Prerequisites
- jupyter notebook,
- the eCommerce dataset,
- python.

#### The Datasets
The dataframe consist of 91,472 entries, with 34 columns. The dataframe is stored in as a pickle file to speed up the storage/retrival and use i.e. merging the datasets into a larger dataset without memory utilization issues.

#### Technology 
Anaconda 

#### Required packages
Numpy, Pandas, Datetime, and Pickle for memory utilization when handling large files without memory utilization issues.

In [1]:
# 1. Import the required packages
import pandas as pd
import numpy as np
import pickle

#### Inspecting the Customer Transaction Dataset
We can load the dataframe as our custome transaction dataset and inspect.

In [2]:
# reload new df
cust_transactions = pd.read_pickle("df_cust_df.pickle")
# inspect
cust_transactions.head(5).T

0  \
customer_id              d5695e900ecca93e1369d850dc1e986b   
zipcode                                           88032.0   
latitude                                       -27.536913   
longitude                                      -48.509018   
city                                        florianopolis   
state                                                  SC   
seller_id                c0aff2da32c17759d30b22fb0af6649d   
order_id                 5c1ae73856bf1d09958ec5ae3a64b73a   
order_status                                    delivered   
date_ordered                          2017-05-11 23:24:53   
date_approved                         2017-05-11 23:35:16   
order_with_courier_date               2017-05-15 09:14:36   
date_delivered                        2017-06-03 09:32:46   
estimated_delivery_date               2017-05-31 00:00:00   
payment_sequence                                        1   
payment_type                                  credit_card   
payment_installments                                    6   
order_value                                        104.28   
item_id                                                 1   
product_id               e6b6e13cf71449a457269f425b89dc74   
expected_shipping_date                2017-05-17 23:35:16   
price                                                89.9   
delivery_cost                                       14.38   
review_id                58273990d50b5e0708eca041f5b0bed4   
review_score                                            1   
review_headline                                       NaN   
review_body                                           NaN   
product_category                                    audio   
mql_id                   f6127362ba4727c54d742b7ab1ae7c1c   
won_date                              2018-04-20 13:24:51   
business_type                                    reseller   
first_contact_date                             2018-04-16   
landing_page             bdaeb752e5c3caa215ebe04c61e22325   
page_origin                                   paid_search   

                                                        1  \
customer_id              d5695e900ecca93e1369d850dc1e986b   
zipcode                                           88032.0   
latitude                                       -27.470734   
longitude                                      -48.468516   
city                                        florianopolis   
state                                                  SC   
seller_id                c0aff2da32c17759d30b22fb0af6649d   
order_id                 5c1ae73856bf1d09958ec5ae3a64b73a   
order_status                                    delivered   
date_ordered                          2017-05-11 23:24:53   
date_approved                         2017-05-11 23:35:16   
order_with_courier_date               2017-05-15 09:14:36   
date_delivered                        2017-06-03 09:32:46   
estimated_delivery_date               2017-05-31 00:00:00   
payment_sequence                                        1   
payment_type                                  credit_card   
payment_installments                                    6   
order_value                                        104.28   
item_id                                                 1   
product_id               e6b6e13cf71449a457269f425b89dc74   
expected_shipping_date                2017-05-17 23:35:16   
price                                                89.9   
delivery_cost                                       14.38   
review_id                58273990d50b5e0708eca041f5b0bed4   
review_score                                            1   
review_headline                                       NaN   
review_body                                           NaN   
product_category                                    audio   
mql_id                   f6127362ba4727c54d742b7ab1ae7c1c   
won_date                              2018-04-20 13:24:51   
business_type                

#### Data Cleaning
We create a copy of the customer geocode dataframe to avoid compromising the data. We expect a transaction to have a unique`customer_id`, `order_id` and `date_ordered`. So we would check the data for duplicates by passing a list of the three parameters to a user defined function for sum of duplicated records `df.duplicated().sum()`. We then print the outcome to obtain the number of duplicates in our records.

In [3]:
# create a copy of data
df = cust_transactions.copy()

In [4]:
# Checking for duplicate records
df_duplicates = (df.duplicated(["customer_id","order_id","date_ordered"]).sum())
# inspect
df_duplicates

91144

We notice that there are 91,144 duplicate records in the transaction dataset. We pass a list of the three parameters to `df.drop_duplicates()` function to drop the duplicates, keeping the `first` of each of the duplicated records.

In [5]:
# drop duplicated records, keeping only the first 
df = (df.drop_duplicates(["customer_id","order_id","date_ordered"], keep ="first"))
# inspect
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 328 entries, 0 to 91226
Data columns (total 34 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              328 non-null    object 
 1   zipcode                  328 non-null    float64
 2   latitude                 328 non-null    float64
 3   longitude                328 non-null    float64
 4   city                     328 non-null    object 
 5   state                    328 non-null    object 
 6   seller_id                328 non-null    object 
 7   order_id                 328 non-null    object 
 8   order_status             328 non-null    object 
 9   date_ordered             328 non-null    object 
 10  date_approved            328 non-null    object 
 11  order_with_courier_date  324 non-null    object 
 12  date_delivered           322 non-null    object 
 13  estimated_delivery_date  328 non-null    object 
 14  payment_sequence        

#### Convert Dates to Datetime Format
We notice that the parameters `date_ordered`, `date_approved`, `order_with_courier_date`, `date_delivered`, `estimated_delivery_date`, `expected_shipping_date`, `won_date`, and `first_contact_date` are not datetime objects. So we convert thems to datetime objects by passing them to the `pd.to_datetime()` function, adding the argument `errors`='coerce' to return NaT (or “not a time”) for any date that cannot be converted. 

In [6]:
# load packages
from datetime import date, datetime, timedelta

In [7]:
# convert dates
df['date_ordered'] = pd.to_datetime(df['date_ordered'], errors='coerce')
df['date_approved'] = pd.to_datetime(df['date_approved'], errors='coerce')
df['order_with_courier_date'] = pd.to_datetime(df['order_with_courier_date'], errors='coerce')
df['date_delivered'] = pd.to_datetime(df['date_delivered'], errors='coerce')
df['estimated_delivery_date'] = pd.to_datetime(df['estimated_delivery_date'], errors='coerce')
df['expected_shipping_date'] = pd.to_datetime(df['expected_shipping_date'], errors='coerce')
df['won_date'] = pd.to_datetime(df['won_date'], errors='coerce')
df['first_contact_date'] = pd.to_datetime(df['first_contact_date'], errors='coerce')
# inspect
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 328 entries, 0 to 91226
Data columns (total 34 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   customer_id              328 non-null    object        
 1   zipcode                  328 non-null    float64       
 2   latitude                 328 non-null    float64       
 3   longitude                328 non-null    float64       
 4   city                     328 non-null    object        
 5   state                    328 non-null    object        
 6   seller_id                328 non-null    object        
 7   order_id                 328 non-null    object        
 8   order_status             328 non-null    object        
 9   date_ordered             328 non-null    datetime64[ns]
 10  date_approved            328 non-null    datetime64[ns]
 11  order_with_courier_date  324 non-null    datetime64[ns]
 12  date_delivered           322 non-n

#### Checking for Missing Data
We apply a user defined function `create missing_values_table()` to determing the presence of any missing data in our dataset. This function calculates the percentage of missing values in the dataset, and return the results in a tabular format. We pass the dataframe to the function to create the table for the missing values. We then call the function to return the table of missing values. 

In [8]:
# define function to to create missing_values_table
def missing_values_table(df):
    """
    Creates a table for missing values.
    """
    # explaing variables
    
    #Total missing values
    mis_val = df.isnull().sum()
    # Percentage of missing values
    mis_val_percent = 100 * df.isnull().sum() / len(df)
        
    # Make a table with the results
    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        
    # Rename the columns
    mis_val_table_ren_columns = mis_val_table.rename(
    columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        
    # Sort the table by percentage of missing descending
    mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
            '% of Total Values', ascending=False).round(1)
        
    # Print some summary information
    print ("The dataframe has " + str(df.shape[1]) + " columns.\n"
        "There are " + str(mis_val_table_ren_columns.shape[0]) +
        " columns that have missing values.")
        
    # Return the dataframe with missing information
    return mis_val_table_ren_columns

In [9]:
# call function to get missing values
df_missing = missing_values_table(df)
# inspect
df_missing 

The dataframe has 34 columns.
There are 6 columns that have missing values.


Missing Values  % of Total Values
review_headline                     296               90.2
review_body                         199               60.7
page_origin                          10                3.0
date_delivered                        6                1.8
order_with_courier_date               4                1.2
business_type                         4                1.2

The outcome reveals that 6 of the of the 34 columns in the dataframe have missing values. We also observe that with the exception of `review_headline` which is not relevant, all the other fields can be used for the modeling. So we proceed to drop `review_headline` from the dataframe. 

In [10]:
# drop unwanted columns
df = df.drop(columns=['review_headline'])
# inspect data
df.head(10).T

0     \
customer_id              d5695e900ecca93e1369d850dc1e986b   
zipcode                                           88032.0   
latitude                                       -27.536913   
longitude                                      -48.509018   
city                                        florianopolis   
state                                                  SC   
seller_id                c0aff2da32c17759d30b22fb0af6649d   
order_id                 5c1ae73856bf1d09958ec5ae3a64b73a   
order_status                                    delivered   
date_ordered                          2017-05-11 23:24:53   
date_approved                         2017-05-11 23:35:16   
order_with_courier_date               2017-05-15 09:14:36   
date_delivered                        2017-06-03 09:32:46   
estimated_delivery_date               2017-05-31 00:00:00   
payment_sequence                                        1   
payment_type                                  credit_card   
payment_installments                                    6   
order_value                                        104.28   
item_id                                                 1   
product_id               e6b6e13cf71449a457269f425b89dc74   
expected_shipping_date                2017-05-17 23:35:16   
price                                                89.9   
delivery_cost                                       14.38   
review_id                58273990d50b5e0708eca041f5b0bed4   
review_score                                            1   
review_body                                           NaN   
product_category                                    audio   
mql_id                   f6127362ba4727c54d742b7ab1ae7c1c   
won_date                              2018-04-20 13:24:51   
business_type                                    reseller   
first_contact_date                    2018-04-16 00:00:00   
landing_page             bdaeb752e5c3caa215ebe04c61e22325   
page_origin                                   paid_search   

                                                     119   \
customer_id              b3c273a2f1c0a7b1e53842aa86de27ad   
zipcode                                            9080.0   
latitude                                       -23.644439   
longitude                                      -46.539885   
city                                          santo andre   
state                                                  SP   
seller_id                f12d3c2a14729ae461b920c11fe20fdc   
order_id                 dadd179aa43944bfe3ebc6e6188b2606   
order_status                                    delivered   
date_ordered                          2017-10-16 17:11:22   
date_approved                         2017-10-16 17:35:19   
order_with_courier_date               2017-10-20 15:51:52   
date_delivered                        2017-10-24 19:50:03   
estimated_delivery_date               2017-10-26 00:00:00   
payment_sequence                                        1   
payment_type                                  credit_card   
payment_installments                                    2   
order_value                                        117.92   
item_id                                                 1   
product_id               e6b6e13cf71449a457269f425b89dc74   
expected_shipping_date                2017-10-20 17:35:19   
price                                               109.9   
delivery_cost                                        8.02   
review_id                ee5cc5a781512a3fb2aec100825adacc   
review_score                                            5   
review_body                                 Tudo em ordem   
product_category                                    audio   
mql_id                   e5c8f30b30415b1fc94d820ba9d4d08c   
won_date                              2018-02-02 13:10:58   
business_type                                    reseller   
first_contact_date                    2018-01-11 00:00:00   
landing_page             b

In [11]:
# inspect bottom 20 of the data
df.tail(20).T

88790  \
customer_id              bab00597e0b0a37f66188dba9063b493   
zipcode                                           20751.0   
latitude                                       -22.886945   
longitude                                      -43.304593   
city                                       rio de janeiro   
state                                                  RJ   
seller_id                698f5fa55a5f73d0740e9d9a773e8093   
order_id                 5a9f638cae9fc9cd16c2f1e748da4c77   
order_status                                    delivered   
date_ordered                          2018-02-24 22:09:07   
date_approved                         2018-02-24 23:00:37   
order_with_courier_date               2018-02-26 18:57:05   
date_delivered                        2018-03-09 20:03:39   
estimated_delivery_date               2018-03-16 00:00:00   
payment_sequence                                        1   
payment_type                                  credit_card   
payment_installments                                    1   
order_value                                         32.09   
item_id                                                 1   
product_id               c9fa6a0dbc5b979d8c99f38c3c9c912f   
expected_shipping_date                2018-02-28 22:31:07   
price                                               17.99   
delivery_cost                                        14.1   
review_id                3e84ede3bcbfa32643759f006a8c7392   
review_score                                            5   
review_body                                           NaN   
product_category                           consoles_games   
mql_id                   9e5523333afdebe2911f20c3e5b452df   
won_date                              2018-03-15 20:31:54   
business_type                                    reseller   
first_contact_date                    2018-03-14 00:00:00   
landing_page             40dec9f3d5259a3d2dbcdab2114fae47   
page_origin                                   paid_search   

                                                                     88900  \
customer_id                               5e00c4a9b6813a22869b75a51b254f18   
zipcode                                                             7113.0   
latitude                                                        -23.461303   
longitude                                                       -46.518866   
city                                                             guarulhos   
state                                                                   SP   
seller_id                                 56e361f411e38dcef17cdc2a3d99628b   
order_id                                  6741e21a24361b07f30063eec3897b37   
order_status                                                     delivered   
date_ordered                                           2018-07-21 16:11:57   
date_approved                                          2018-07-21 16:24:08   
order_with_courier_date                                2018-07-23 15:00:00   
date_delivered                                         2018-07-27 15:14:31   
estimated_delivery_date                                2018-08-07 00:00:00   
payment_sequence                                                         1   
payment_type                                                   credit_card   
payment_installments                                                     1   
order_value                                                          98.65   
item_id                                                                  1   
product_id                                19a6ffd4e245a92747a693805163a60f   
expected_shipping_date                                 2018-07-24 16:24:08   
price                                                                79.99   
delivery_cost                                                        18.66   
review_id                                 66a93aaaec9be9ecf385c8e9ee6280ad   
review_score               

In [12]:
# inspect data info
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 328 entries, 0 to 91226
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   customer_id              328 non-null    object        
 1   zipcode                  328 non-null    float64       
 2   latitude                 328 non-null    float64       
 3   longitude                328 non-null    float64       
 4   city                     328 non-null    object        
 5   state                    328 non-null    object        
 6   seller_id                328 non-null    object        
 7   order_id                 328 non-null    object        
 8   order_status             328 non-null    object        
 9   date_ordered             328 non-null    datetime64[ns]
 10  date_approved            328 non-null    datetime64[ns]
 11  order_with_courier_date  324 non-null    datetime64[ns]
 12  date_delivered           322 non-n

#### Filling Missing Values
We proceed to fill the missing values but before then we investigate the dataframe to find unique values in each of the columns/fields. We achieve this by looping (for loop) the columns in the dataframe then printing the unique values by passing the columns to `df.unique()` function.

In [13]:
for col in df:
  print(df[col].unique())

['d5695e900ecca93e1369d850dc1e986b' 'b3c273a2f1c0a7b1e53842aa86de27ad'
 'c5912afa747c4f74458aecee49c8f8f2' '1145efd0cc4b891ba7df8d75f1530ec0'
 '96884edb162028aa1d676fdd30629d7b' 'f02281ff50731e8a5e07bbc0a45d4912'
 '4999b83ecd2fab104b958ce8d13714cf' 'aeb07999154eeee6ed241aa2b3137c99'
 '03a17587335e8de93ffc1cae3a515cab' 'bf06469eadee11e635909ecabf966022'
 'a562db3c7cb9a68947debd30879b491e' '8f0b26345ceb1dd629eaecad15b37e45'
 '6f5ab7951708c5ecf54e051674e7a419' '0efd489ae0fdcbb706013c9b12ece141'
 '4dc755960dfc89677457aa98e0e4d4b1' '3f244eb4a15e626a8fee14bc089303c9'
 'b9c32ffb4932ea3b8ee7c7ccf6a1c55c' '65c9e0808cb89c7417c7ed4ad2d78fe3'
 '5e320e6a4f03e2e74bc22e1158237fbe' 'a26e57e9f234d61b47047662377e8772'
 'b7fb62f14c17de6c1cb77ed5330dc869' 'aeec4d2f2951cc7132dd0a1186df6bb3'
 '072a4d5c91be7fd81a21c401668a27a0' '7cc29bd694a82357723df9961ab7d3a6'
 '1dc9101f2d3193e393d256dcfb50b527' '06665c03fe91dc717e53e8316090ca07'
 '8ddad16131b6e876e6b5e1c8d01f14da' 'c82db19ec1f02359a3ba17d5dca4290a'
 'ba17

['bdaeb752e5c3caa215ebe04c61e22325' 'b76ef37428e6799c421989521c0e5077'
 '0d6bc3c00e4e64927cae2e8d9c6a0b9b' '35c9b150ab36fe584c1f24fd458c453a'
 '22c29808c4f815213303f8933030604c' '88740e65d5d6b056e0cda098e1ea6313'
 '65d9f9d71e562365e8b44037c2888d98' '1ceb590cd1e00c7ee95220971f82693d'
 'b634a925e3768734c5b7b2325c88a082' '30077c17f2ec5010a82e37ad8925b95f'
 'a7982125ff7aa3b2054c6e44f9d28522' '1baca9df5850e5ba75545d8b2eb8d966'
 '68f817c510c978e3a3a7474ad2aeb75c' '0f7f8ed9a1db8fc2a42e273972db10d4'
 'a0fbc0263cb98bdb0d72b9f9f283a172' 'b6885f18d203a61176418c1fb3764815'
 'b16779a74f2b20eb9a17a785519f63a0' '7fa6214d82e911d070f51ef79381b956'
 'ce1a65abd0973638f1c887a6efcfa82d' 'd83b0d0e48c8447d1d5507a44027a955'
 'fbc24da54d531c6204ae2d17b1090bb1' '40dec9f3d5259a3d2dbcdab2114fae47'
 '0218f6be0b76aca72ab4d00ee9e8cf10' '4e82dd1f6d00626bda0723eef0a269a6'
 'e492ee5eaf1697716985cc6f33f9cd9b' '358adb2fee9a122549618e46925a00a5'
 '649b7b8b46b254497e92038d88f6ced8' '58326e62183c14b0c03085c33b9fdc44'
 '73f3

In [14]:
# create new DataFrame that only contains rows with NaNs in page_origin column
df_page_origin_nans = df.loc[df['page_origin'].isnull()]
#view results
df_page_origin_nans

customer_id  zipcode   latitude  longitude  \
17708  5cf67e8b0f3046587f30daa2ae1c3e69  35700.0 -19.463613 -44.244414   
18386  67f17583aa3d40a008f98ff8abc14069  35700.0 -19.463613 -44.244414   
60307  f0a67330d461ab7c99fbb328e33cf54a  80060.0 -25.432111 -49.259768   
60588  bfecf4e7164c2af64001ea4bed8eea1e  80060.0 -25.432111 -49.259768   
60869  7d9692e3a7850c9ce64a089b62a910da  80060.0 -25.432111 -49.259768   
61150  7d9c3f29dad31b17f13fd3a928ea4955  80060.0 -25.432111 -49.259768   
61431  6ed3a6f46ef83d356e7152ac4a4388db  80060.0 -25.432111 -49.259768   
81780  833d52f3b53923c7f3e2371dced43ce3  35700.0 -19.463613 -44.244414   
83136  4d719448b1cc626699bc3bbbdbbc0bb7  35700.0 -19.463613 -44.244414   
85125  90f85ef7f6b7f07be42e73c4b8f298b9  80060.0 -25.432111 -49.259768   

              city state                         seller_id  \
17708  sete lagoas    MG  213fafb0ca06fb3d5886579c2565791b   
18386  sete lagoas    MG  213fafb0ca06fb3d5886579c2565791b   
60307     curitiba    PR  0cab2da43793a6f3c5ed8514c5f54627   
60588     curitiba    PR  0cab2da43793a6f3c5ed8514c5f54627   
60869     curitiba    PR  0cab2da43793a6f3c5ed8514c5f54627   
61150     curitiba    PR  0cab2da43793a6f3c5ed8514c5f54627   
61431     curitiba    PR  0cab2da43793a6f3c5ed8514c5f54627   
81780  sete lagoas    MG  213fafb0ca06fb3d5886579c2565791b   
83136  sete lagoas    MG  213fafb0ca06fb3d5886579c2565791b   
85125     curitiba    PR  0cab2da43793a6f3c5ed8514c5f54627   

                               order_id order_status        date_ordered  ...  \
17708  be49c4e4b3b0f04ec7752ea004feec8c    delivered 2018-08-05 14:07:10  ...   
18386  c6c0dc0b58d87603f5b07c6283355d5c    delivered 2017-09-20 17:16:44  ...   
60307  f14ffd28776719a57c5bbf16cce58dd4    delivered 2017-07-06 11:13:08  ...   
60588  2bb4f4b23284b2f435007ab797c29edc    delivered 2018-08-13 13:42:52  ...   
60869  2067e6809c31a615c37a470db2e6bf89    delivered 2018-06-21 14:56:17  ...   
61150  0d123aa33c3954b2685ef6807ffc9bc0    delivered 2017-09-07 09:32:18  ...   
61431  e6a7754d776e95d1a2b8e6d6f2d7ac9d    delivered 2017-04-25 19:35:54  ...   
81780  9298c4a7ed6bd4d2cc8433f7e4053fe5    delivered 2017-10-30 22:36:14  ...   
83136  28a9bfa90eb24efe91bfc50b7cc1b4d1    delivered 2017-10-16 08:07:41  ...   
85125  7ccdf2db4cb80263b5535a6cfad598e8    delivered 2018-03-05 22:26:48  ...   

                              review_id review_score  \
17708  8625f35712c54f5b27e7eb76535b1089            4   
18386  8c264c02db8c365453f1995ff8a5a39a            5   
60307  e79e184ed8d1b2aebaa5ade805f60400            4   
60588  43ab7facdcfbe5763045a2b1f4d7fb31            5   
60869  c25b4044534dd12c0f9238799e1b1e5f            5   
61150  945206c1be3da6a3a0344ff00af84853            5   
61431  3abf871d0ada72f1a556f24bf0e0c275            3   
81780  cc8307450b5acf9ebe480f9f27215e38            3   
83136  183ef5f57ae1676eb824da9d018f9812            5   
85125  5fa6d429e4bf560ca9786763f6247dcb            1   

                                             review_body product_category  \
17708  o produto foi entregue conforme combinado. Fun...         pet_shop   
18386                                                NaN         pet_shop   
60307                                                NaN       cool_stuff   
60588                                                NaN       cool_stuff   
60869       PRODUTO BOM, ENTREGA ANTES DO PRAZO PREVISTO       cool_stuff   
61150                                                NaN       cool_stuff   
61431                                                NaN       cool_stuff   
81780                                                NaN   consoles_games   
83136  Sim recomendaria pq alem de vir correto veio a...   consoles_games   
85125  Produto não chegou no prazo, que era há quase ...   consoles_games   

                                 mql_id            won_date  business_type  \
17708  1785cff273aedbd875c814b46928ad7f 2018-05-17 18:11:29       reseller   
18386  1785cff273

First we consider the column `page_origin` which has 10 missing values. The `page_origin` refers to the page (e.g. social, email, etc) through which the marketing team persuade the page visitors to take one specific action, that is respond to the marketing campaigns. Due to the large number of unique values in the entire dataframe, we first attempt to find the number of unique values in the `page_origin` by passing the column name to the `df.unique()` function. 

In [15]:
# Find Unique Values in One Column
df_unique_po = df.page_origin.unique()
# inspect
df_unique_po

array(['paid_search', 'organic_search', 'unknown', 'social', 'email',
       'referral', nan, 'direct_traffic', 'display', 'other'],
      dtype=object)

We notice that the clients typically arrive at `landing_pages` via the `page_origion` via actions within the advertising campaign. The actions se includes paid_search, social media posts, email, to mention a few. We would attempt to fill the missing/unknown values in the `page_origin` column using values in the `landing_page` column. 

#### Fill Missing Values Based on Another Column
We notice from the unique `page_origin` data above that in addition to the `nan`, the column also have had values which are `unknown`. This suggests that some clients arrived at the `landing_page` through unspecified sources. We would attempt to fill the missing/unknown values (i.e. the `nan` and `unknown`) based on the values in the `landing_page` column. 

We begin with the `NaN`, i.e. where the `page_origin` == `NaN`. We would attempt to use enteries in the `landing_page` field to do a lookup for matiching the `landing_page` and the `page_origin`. Once this is found, `NaN` would be replaced by the `page_origin` name. 

In [17]:
# mapping for nan
mapping = (df.query('page_origin != "NaN"')
             .drop_duplicates('landing_page')
             .set_index('landing_page')
             .page_origin)

Now, pass `mapping` to the `df.map()` function then append a `.fillna(NaN)` to fill the `NaN` values in the `page_origin`.

In [18]:
# fill the nan values
df['page_origin'] = df['landing_page'].map(mapping).fillna('NaN')

We inspect the outcome by creating a datafran for the rows with `NaN` in the `page_origin` column.

In [19]:
# inspect
# create new DataFrame that only contains rows with NaNs in page_origin column
df_page_origin_nans = df.loc[df['page_origin'].isnull()]
#view results
df_page_origin_nans

Empty DataFrame
Columns: [customer_id, zipcode, latitude, longitude, city, state, seller_id, order_id, order_status, date_ordered, date_approved, order_with_courier_date, date_delivered, estimated_delivery_date, payment_sequence, payment_type, payment_installments, order_value, item_id, product_id, expected_shipping_date, price, delivery_cost, review_id, review_score, review_body, product_category, mql_id, won_date, business_type, first_contact_date, landing_page, page_origin]
Index: []

[0 rows x 33 columns]

We repeate the process for the `page_origin` with values entered as `unknown`, that is `page_origin` == `unknown`.

In [21]:
# create new DataFrame that only contains rows with NaNs in page_origin column
df_page_origin_unknowns = df.loc[df['page_origin'] == 'unknown']
#view results
df_page_origin_unknowns

customer_id  zipcode   latitude  longitude  \
3038   a562db3c7cb9a68947debd30879b491e  14802.0 -21.760806 -48.172285   
8074   7b3a2b1682c37d9df9e836d397b963a1  11015.0 -23.959748 -46.313592   
8410   e4c44f8397c63555e4d0b11f4e7addc5  11015.0 -23.959748 -46.313592   
16870  74923e29264de38aca1e2515fd8e5015   8710.0 -23.519375 -46.192690   
21385  7c29841bc24b6a0a94f8f6a9769fb54c  15025.0 -20.802436 -49.395624   
23406  fc67e04362ada373e1f7e7b6dab539e8  37410.0 -21.692168 -45.258049   
28819  a875682b67eb7d30c5fdd0408711e669  13330.0 -23.089035 -47.214419   
29064  1700391c8dfc8bd005f553a061675982  13330.0 -23.089035 -47.214419   
30827  0faf52fb32003ad8ca5853afe236ac71   5372.0 -23.583166 -46.761244   
31530  4cd281d8030921fc1b77650f63efdda3  37540.0 -22.243203 -45.713303   
31897  2d76ccbc1cecea0ffef654de59923d87  37540.0 -22.243203 -45.713303   
33179  7d2af39c5453e92ad7baad03e370f1d7  85905.0 -24.704651 -53.737053   
33253  137105b0420e2f6a83fe51bd237888b0  14802.0 -21.760806 -48.172285   
34825  7227f6bc1ad4dd7ea6443ea71ff722ae   4537.0 -23.594499 -46.677721   
34947  551ea6ba7a69031489911338c8999014   4537.0 -23.594499 -46.677721   
35069  61ed37058a9365674a8cb35568941b0d  97050.0 -29.691002 -53.806237   
35326  e20e7de2c0080cce2dbdef34344ce227  11015.0 -23.959748 -46.313592   
47076  bc80b4e6fc53eccd155fd449dbefda27  13418.0 -22.728804 -47.631738   
48366  1a7b30b740fef8190f7de1e678ce6290   9070.0 -23.653212 -46.554463   
50668  399ea879d045ade57e0a91c195b73aba   3694.0 -23.537922 -46.477696   
50780  eba2b167e764f26251de139177a73708  17209.0 -22.275687 -48.542600   
50888  bc7739e799fb6568ea343b5616d23e62  17209.0 -22.275687 -48.542600   
53016  82829411038c21f87c23d8a785ed1ead   7152.0 -23.408025 -46.471333   
55269  21249ce919ebfa3728a482f7154fbe39  15025.0 -20.802436 -49.395624   
65021  a8dddbb164f3848ad6a58dace214ed7b   3210.0 -23.601765 -46.547851   
65175  99b3b3709b00a7b817833b1be4ff0729  87015.0 -23.428806 -51.971831   
65543  8ebae1632217e6669b1a563f319d238e  95840.0 -29.935889 -51.721173   
69242  0b4442cad8eab4b19f6f4bd2132846a0   7180.0 -23.453118 -46.474442   
69303  cd0b86c73331d47ddbbc5e6b6ab0a2b7   7180.0 -23.453118 -46.474442   
70569  84df5ad0a2f0929761a5d55a1662cc1d  20751.0 -22.886945 -43.304593   
70679  e40e7edfb1c1a656dce58946a597df29  81280.0 -25.434992 -49.352495   
71947  18e6b781ea6f01f2b0a73bfc54aa2509  30668.0 -20.009133 -44.029840   
72839  87a4ee899da050ef80de6de979521e12  81200.0 -25.440088 -49.345266   
73056  76ed00bb91c35675557f53cf791e6f04  81200.0 -25.440088 -49.345266   
73273  bdd1804f10b8848c98659a9034876414  81200.0 -25.440088 -49.345266   
75253  b94a28d76c3a1186c1d20f6b23d5f6f7  88037.0 -27.600663 -48.514121   
76355  e5ccfd48cc43aceb9d93703b68d8aad0  88870.0 -28.348012 -49.300554   
76436  3306b138f542d4f9ad7d9b59ffe3b83e  23943.0 -22.905485 -44.319284   
76745  1bf918ef7f7109ab40fce559804b274e   4537.0 -23.594499 -46.677721   
76867  70aecb98ec965a174733a3021c64373b  97050.0 -29.691002 -53.806237   
80484  86911781ab615c91b809029803289020   6852.0 -23.711999 -46.859312   
81683  0491083e2ce781339e49382f4ccee438   7152.0 -23.408025 -46.471333   
86975  a31e61974efff601345e65e5d6b7ff31  13330.0 -23.089035 -47.214419   
87955  0ad854ccd5f47d7c60dc5f0312901d6d  37480.0 -21.974109 -45.347956   
88790  bab00597e0b0a37f66188dba9063b493  20751.0 -22.886945 -43.304593   
90121  2d00fe84cd1a1a9f9477fb93510f323a  13097.0 -22.846952 -47.027434   
90712  44531f7ae52e2312ab3ca7f58778b41f   8710.0 -23.519375 -46.192690   
91081  953b4d3b6f785d9148b934d68125a74f  88037.0 -27.600663 -48.514121   

                        city state                         seller_id  \
3038              araraquara    SP  af3ef48d0e13835e529c29ac573c63e5   
8074                  santos    SP  ed8cb7b190ceb6067227478e48cf8dde   
8410                  santos    SP  ed8cb7b190ceb6067227478e48cf8dde   
16870        mogi das cruzes    SP  596849622429351f47b32e6cae1055ff   
21385  sao jose do rio preto    SP  cc63f

In [23]:
# mapping for unknown
mapping = (df.query('page_origin != "unknown"')
             .drop_duplicates('landing_page')
             .set_index('landing_page')
             .page_origin)

In [24]:
# fill unmapped values
df['page_origin'] = df['landing_page'].map(mapping).fillna('unknown')

In [25]:
# inspect
# create new DataFrame that only contains rows with unknown in page_origin column
df_page_origin_unknowns = df.loc[df['page_origin'] == 'unknown']
#view results
df_page_origin_unknowns

customer_id  zipcode   latitude  longitude  \
3038   a562db3c7cb9a68947debd30879b491e  14802.0 -21.760806 -48.172285   
8074   7b3a2b1682c37d9df9e836d397b963a1  11015.0 -23.959748 -46.313592   
8410   e4c44f8397c63555e4d0b11f4e7addc5  11015.0 -23.959748 -46.313592   
16870  74923e29264de38aca1e2515fd8e5015   8710.0 -23.519375 -46.192690   
21385  7c29841bc24b6a0a94f8f6a9769fb54c  15025.0 -20.802436 -49.395624   
23406  fc67e04362ada373e1f7e7b6dab539e8  37410.0 -21.692168 -45.258049   
28819  a875682b67eb7d30c5fdd0408711e669  13330.0 -23.089035 -47.214419   
29064  1700391c8dfc8bd005f553a061675982  13330.0 -23.089035 -47.214419   
30827  0faf52fb32003ad8ca5853afe236ac71   5372.0 -23.583166 -46.761244   
31530  4cd281d8030921fc1b77650f63efdda3  37540.0 -22.243203 -45.713303   
31897  2d76ccbc1cecea0ffef654de59923d87  37540.0 -22.243203 -45.713303   
33179  7d2af39c5453e92ad7baad03e370f1d7  85905.0 -24.704651 -53.737053   
33253  137105b0420e2f6a83fe51bd237888b0  14802.0 -21.760806 -48.172285   
34825  7227f6bc1ad4dd7ea6443ea71ff722ae   4537.0 -23.594499 -46.677721   
34947  551ea6ba7a69031489911338c8999014   4537.0 -23.594499 -46.677721   
35069  61ed37058a9365674a8cb35568941b0d  97050.0 -29.691002 -53.806237   
35326  e20e7de2c0080cce2dbdef34344ce227  11015.0 -23.959748 -46.313592   
47076  bc80b4e6fc53eccd155fd449dbefda27  13418.0 -22.728804 -47.631738   
48366  1a7b30b740fef8190f7de1e678ce6290   9070.0 -23.653212 -46.554463   
50668  399ea879d045ade57e0a91c195b73aba   3694.0 -23.537922 -46.477696   
50780  eba2b167e764f26251de139177a73708  17209.0 -22.275687 -48.542600   
50888  bc7739e799fb6568ea343b5616d23e62  17209.0 -22.275687 -48.542600   
53016  82829411038c21f87c23d8a785ed1ead   7152.0 -23.408025 -46.471333   
55269  21249ce919ebfa3728a482f7154fbe39  15025.0 -20.802436 -49.395624   
65021  a8dddbb164f3848ad6a58dace214ed7b   3210.0 -23.601765 -46.547851   
65175  99b3b3709b00a7b817833b1be4ff0729  87015.0 -23.428806 -51.971831   
65543  8ebae1632217e6669b1a563f319d238e  95840.0 -29.935889 -51.721173   
69242  0b4442cad8eab4b19f6f4bd2132846a0   7180.0 -23.453118 -46.474442   
69303  cd0b86c73331d47ddbbc5e6b6ab0a2b7   7180.0 -23.453118 -46.474442   
70569  84df5ad0a2f0929761a5d55a1662cc1d  20751.0 -22.886945 -43.304593   
70679  e40e7edfb1c1a656dce58946a597df29  81280.0 -25.434992 -49.352495   
71947  18e6b781ea6f01f2b0a73bfc54aa2509  30668.0 -20.009133 -44.029840   
72839  87a4ee899da050ef80de6de979521e12  81200.0 -25.440088 -49.345266   
73056  76ed00bb91c35675557f53cf791e6f04  81200.0 -25.440088 -49.345266   
73273  bdd1804f10b8848c98659a9034876414  81200.0 -25.440088 -49.345266   
75253  b94a28d76c3a1186c1d20f6b23d5f6f7  88037.0 -27.600663 -48.514121   
76355  e5ccfd48cc43aceb9d93703b68d8aad0  88870.0 -28.348012 -49.300554   
76436  3306b138f542d4f9ad7d9b59ffe3b83e  23943.0 -22.905485 -44.319284   
76745  1bf918ef7f7109ab40fce559804b274e   4537.0 -23.594499 -46.677721   
76867  70aecb98ec965a174733a3021c64373b  97050.0 -29.691002 -53.806237   
80484  86911781ab615c91b809029803289020   6852.0 -23.711999 -46.859312   
81683  0491083e2ce781339e49382f4ccee438   7152.0 -23.408025 -46.471333   
86975  a31e61974efff601345e65e5d6b7ff31  13330.0 -23.089035 -47.214419   
87955  0ad854ccd5f47d7c60dc5f0312901d6d  37480.0 -21.974109 -45.347956   
88790  bab00597e0b0a37f66188dba9063b493  20751.0 -22.886945 -43.304593   
90121  2d00fe84cd1a1a9f9477fb93510f323a  13097.0 -22.846952 -47.027434   
90712  44531f7ae52e2312ab3ca7f58778b41f   8710.0 -23.519375 -46.192690   
91081  953b4d3b6f785d9148b934d68125a74f  88037.0 -27.600663 -48.514121   

                        city state                         seller_id  \
3038              araraquara    SP  af3ef48d0e13835e529c29ac573c63e5   
8074                  santos    SP  ed8cb7b190ceb6067227478e48cf8dde   
8410                  santos    SP  ed8cb7b190ceb6067227478e48cf8dde   
16870        mogi das cruzes    SP  596849622429351f47b32e6cae1055ff   
21385  sao jose do rio preto    SP  cc63f

In [26]:
# inspect
# create new DataFrame that only contains rows with unknown in page_origin column
df_page_origin_unknowns = df.loc[df['page_origin'] == 'unknown']
#view results
df_page_origin_unknowns

customer_id  zipcode   latitude  longitude  \
3038   a562db3c7cb9a68947debd30879b491e  14802.0 -21.760806 -48.172285   
8074   7b3a2b1682c37d9df9e836d397b963a1  11015.0 -23.959748 -46.313592   
8410   e4c44f8397c63555e4d0b11f4e7addc5  11015.0 -23.959748 -46.313592   
16870  74923e29264de38aca1e2515fd8e5015   8710.0 -23.519375 -46.192690   
21385  7c29841bc24b6a0a94f8f6a9769fb54c  15025.0 -20.802436 -49.395624   
23406  fc67e04362ada373e1f7e7b6dab539e8  37410.0 -21.692168 -45.258049   
28819  a875682b67eb7d30c5fdd0408711e669  13330.0 -23.089035 -47.214419   
29064  1700391c8dfc8bd005f553a061675982  13330.0 -23.089035 -47.214419   
30827  0faf52fb32003ad8ca5853afe236ac71   5372.0 -23.583166 -46.761244   
31530  4cd281d8030921fc1b77650f63efdda3  37540.0 -22.243203 -45.713303   
31897  2d76ccbc1cecea0ffef654de59923d87  37540.0 -22.243203 -45.713303   
33179  7d2af39c5453e92ad7baad03e370f1d7  85905.0 -24.704651 -53.737053   
33253  137105b0420e2f6a83fe51bd237888b0  14802.0 -21.760806 -48.172285   
34825  7227f6bc1ad4dd7ea6443ea71ff722ae   4537.0 -23.594499 -46.677721   
34947  551ea6ba7a69031489911338c8999014   4537.0 -23.594499 -46.677721   
35069  61ed37058a9365674a8cb35568941b0d  97050.0 -29.691002 -53.806237   
35326  e20e7de2c0080cce2dbdef34344ce227  11015.0 -23.959748 -46.313592   
47076  bc80b4e6fc53eccd155fd449dbefda27  13418.0 -22.728804 -47.631738   
48366  1a7b30b740fef8190f7de1e678ce6290   9070.0 -23.653212 -46.554463   
50668  399ea879d045ade57e0a91c195b73aba   3694.0 -23.537922 -46.477696   
50780  eba2b167e764f26251de139177a73708  17209.0 -22.275687 -48.542600   
50888  bc7739e799fb6568ea343b5616d23e62  17209.0 -22.275687 -48.542600   
53016  82829411038c21f87c23d8a785ed1ead   7152.0 -23.408025 -46.471333   
55269  21249ce919ebfa3728a482f7154fbe39  15025.0 -20.802436 -49.395624   
65021  a8dddbb164f3848ad6a58dace214ed7b   3210.0 -23.601765 -46.547851   
65175  99b3b3709b00a7b817833b1be4ff0729  87015.0 -23.428806 -51.971831   
65543  8ebae1632217e6669b1a563f319d238e  95840.0 -29.935889 -51.721173   
69242  0b4442cad8eab4b19f6f4bd2132846a0   7180.0 -23.453118 -46.474442   
69303  cd0b86c73331d47ddbbc5e6b6ab0a2b7   7180.0 -23.453118 -46.474442   
70569  84df5ad0a2f0929761a5d55a1662cc1d  20751.0 -22.886945 -43.304593   
70679  e40e7edfb1c1a656dce58946a597df29  81280.0 -25.434992 -49.352495   
71947  18e6b781ea6f01f2b0a73bfc54aa2509  30668.0 -20.009133 -44.029840   
72839  87a4ee899da050ef80de6de979521e12  81200.0 -25.440088 -49.345266   
73056  76ed00bb91c35675557f53cf791e6f04  81200.0 -25.440088 -49.345266   
73273  bdd1804f10b8848c98659a9034876414  81200.0 -25.440088 -49.345266   
75253  b94a28d76c3a1186c1d20f6b23d5f6f7  88037.0 -27.600663 -48.514121   
76355  e5ccfd48cc43aceb9d93703b68d8aad0  88870.0 -28.348012 -49.300554   
76436  3306b138f542d4f9ad7d9b59ffe3b83e  23943.0 -22.905485 -44.319284   
76745  1bf918ef7f7109ab40fce559804b274e   4537.0 -23.594499 -46.677721   
76867  70aecb98ec965a174733a3021c64373b  97050.0 -29.691002 -53.806237   
80484  86911781ab615c91b809029803289020   6852.0 -23.711999 -46.859312   
81683  0491083e2ce781339e49382f4ccee438   7152.0 -23.408025 -46.471333   
86975  a31e61974efff601345e65e5d6b7ff31  13330.0 -23.089035 -47.214419   
87955  0ad854ccd5f47d7c60dc5f0312901d6d  37480.0 -21.974109 -45.347956   
88790  bab00597e0b0a37f66188dba9063b493  20751.0 -22.886945 -43.304593   
90121  2d00fe84cd1a1a9f9477fb93510f323a  13097.0 -22.846952 -47.027434   
90712  44531f7ae52e2312ab3ca7f58778b41f   8710.0 -23.519375 -46.192690   
91081  953b4d3b6f785d9148b934d68125a74f  88037.0 -27.600663 -48.514121   

                        city state                         seller_id  \
3038              araraquara    SP  af3ef48d0e13835e529c29ac573c63e5   
8074                  santos    SP  ed8cb7b190ceb6067227478e48cf8dde   
8410                  santos    SP  ed8cb7b190ceb6067227478e48cf8dde   
16870        mogi das cruzes    SP  596849622429351f47b32e6cae1055ff   
21385  sao jose do rio preto    SP  cc63f

In [29]:
df_page_origin_unknowns = df.page_origin[df.page_origin == 'unknown'].count()

In [30]:
df_page_origin_unknowns

48

We notice that the page origin are not filled using the mapping method. This is because a page origin is unknown if the page is not accessible by the users (in this case the matketing team). Thus the ad listing is not available and therefore stop showing in the page listings. Until the issue is resolved by identify and fixing the error to make sure the page origin is accessible from any device, location, or browser we cannot fill in the page values.

#### George Jordan Thursday, April 20th 2023